lib/AST/ASTDumper.cpp

```diff
@@ -2723,8 +2731,9 @@ void Expr::dump() const {
 void Expr::dump(raw_ostream &OS,
                  llvm::function_ref<Type(const Expr *)> getTypeOfExpr,
                  llvm::function_ref<Type(const TypeLoc &)> getTypeOfTypeLoc,
+                 llvm::function_ref<Type(const KeyPathExpr *E, unsigned index)> getTypeOfKeyPathComponent,
```

Comment:
Rather than scattering `function_ref`s like this throughout Expr.h, we should refactor to add some kind of `ExprTypeMap` type with these kinds of operations. The base class would just call through to the AST node, but `ConstraintSystem` could derive from it and override them to look into its internal tables. Then we should rip out all the `getType`, `setType`, etc. closures we pass around and use that instead.

That's a bigger job than I'm prepared to do in this PR, though.

Reply:
That's fair!